pkg/ruler/api.go

```diff
@@ -482,17 +484,20 @@ func (a *API) CreateRuleGroup(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	rgs, err := a.store.ListRuleGroupsForUserAndNamespace(req.Context(), userID, "")
-	if err != nil {
-		level.Error(logger).Log("msg", "unable to fetch current rule groups for validation", "err", err.Error(), "user", userID)
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
+	// Only fetch rule groups when enforcing a max number of groups for this tenant.
+	if rgsLimit := a.limits.RulerMaxRuleGroupsPerTenant(userID); rgsLimit > 0 {
+		rgs, err := a.store.ListRuleGroupsForUserAndNamespace(req.Context(), userID, "")
```

Comment:
In `ListRuleGroupsForUserAndNamespace` there's more than meets the eye. 

In the PR description, you've stated

>  we load every existing rule group for that tenant and count them

But correct me if I'm wrong, I don't think this is correct. We never "load" the rule groups. In `ListRuleGroupsForUserAndNamespace` it uses `userBucket.Iter` which does a listing of the objects by prefix, in my experience this is extremely fast so I'm surprised to hear that this operation is what's slowing things down.